contracts/cw-ics721-bridge/src/contract.rs

```diff
@@ -0,0 +1,451 @@
+    // (2) Store each NFT we receive in our maps (giving it a classID
+    //     of its own address) and do the same lookup / transfer
+    //     regardless of if the NFT is becoming a native one.
+    //
+    // To avoid special logic when receiving packets, we go the save
+    // route. We can change this back later if we'd like.
+    if !NFT_CONTRACT_TO_CLASS_ID.has(deps.storage, info.sender.clone()) {
+        // This is a new NFT so we give it a "classID". If the map
+        // does have the key, it means that either (a) we have seen
+        // the contract before, or (b) this is a NFT that we received
+        // over IBC and made a new contract for.
+        NFT_CONTRACT_TO_CLASS_ID.save(
+            deps.storage,
+            info.sender.clone(),
+            &info.sender.to_string(),
+        )?;
+        CLASS_ID_TO_NFT_CONTRACT.save(deps.storage, info.sender.to_string(), &info.sender)?;
+    }
+
+    // Class ID is the IBCd ID, or the contract address.
+    let class_id = NFT_CONTRACT_TO_CLASS_ID.load(deps.storage, info.sender.clone())?;
+
+    // Can't allow specifying the class URI in the message as this
+    // could cause multiple different class IDs to be submitted for
+    // the same NFT collection by users. No way to decide on 'correct'
+    // one.
+    let class_uri = CLASS_ID_TO_CLASS_URI
+        .may_load(deps.storage, class_id.clone())?
+        .flatten();
+
+    let UniversalNftInfoResponse { token_uri, .. } = deps.querier.query_wasm_smart(
+        info.sender.clone(),
+        &cw721::Cw721QueryMsg::NftInfo {
+            token_id: token_id.clone(),
+        },
+    )?;
+
+    let ibc_message = NonFungibleTokenPacketData {
+        classId: class_id.clone(),
+        classUri: class_uri,
+        tokenIds: vec![token_id.clone()],
+        tokenUris: vec![token_uri.unwrap_or_default()], // Currently token_uri is optional in cw721 - we set to empty string as default.
+        sender: sender.into_string(),
+        receiver: msg.receiver,
+    };
+    let ibc_message = IbcMsg::SendPacket {
+        channel_id: msg.channel_id.clone(),
+        data: to_binary(&ibc_message)?,
+        timeout: msg.timeout,
+    };
+
+    // Transfer message to send NFT to escrow for channel.
+    let channel_escrow = CHANNELS.load(deps.storage, msg.channel_id.clone())?;
+
+    let transfer_message = cw721::Cw721ExecuteMsg::TransferNft {
+        recipient: channel_escrow.to_string(),
+        token_id: token_id.clone(),
+    };
+    let transfer_message = WasmMsg::Execute {
+        contract_addr: info.sender.into_string(),
+        msg: to_binary(&transfer_message)?,
+        funds: vec![],
+    };
+
+    Ok(Response::default()
+        .add_attribute("method", "execute_receive_nft")
+        .add_attribute("token_id", token_id)
+        .add_attribute("class_id", class_id)
+        .add_attribute("escrow", channel_escrow)
+        .add_attribute("channel_id", msg.channel_id)
+        .add_message(transfer_message)
```

Comment:
noob question, what's the difference between "add message" and "add submessage"?

Reply:
submessage means we can handle the message results e.g. if it fails or works. messages are just blind fired AKA we do not care about the result

Reply:
So for sub message we get to catch them in a Reply function
For message we they just arrive based on whatever type of message we constructed?

Reply:
> So for sub message we get to catch them in a Reply function
Yes, this way we can handle the result. I.E. get the address of an instantiate contract or handle an error.

More info on Messages and SubMessages in the docs: https://docs.cosmwasm.com/docs/1.0/smart-contracts/message/submessage